ipa-core/src/utils/arraychunks.rs

```diff
@@ -17,17 +17,17 @@ pub struct ArrayChunk<I, const L: usize> {
     iter: I,
 }
 
-impl<I: Iterator<Item = Fp61BitPrime>> ArrayChunkIterator for I {}
+impl<F: PrimeField, I: Iterator<Item = F>> ArrayChunkIterator for I {}
 
 #[allow(clippy::while_let_on_iterator)]
-impl<I, const L: usize> Iterator for ArrayChunk<I, L>
+impl<F: PrimeField, I, const L: usize> Iterator for ArrayChunk<I, L>
```

Comment:
Are you sure that `Iterator` is appropriate here? you never return more than one element

Reply:
I don't understand this comment. Is there a better alternative? The reason why I implement this is that I want to convert an iterator over items of type F to an iterator over items [F; L], i.e. I want to have L elements at a time. I don't know how else to do it. With implementing iterator I can have a customized next function that pulls L elements at a time from the underlying iterator of items of type F. @akoshelev please let me know your thoughts, I am happy to change it in a follow up pr

Reply:
So I looked at the code again and you're right - this does return more than one value. And yes, opaque types require you to do the boxing. 
The recursion implementation is still a bit confusing for me. What is the goal for you here? The state machine that you build in runtime for iterators that point to other iterators on the heap can be quite large and I don't think it is cheaper/faster to do than collecting your input into a vector and then compressing it using your algorithm.


My understanding of the protocol here, which could be completely wrong, is that you have a collection of field values on your input of some size $N$. Your goal is to compress it into a vector of size no more than  λ. You do that by inspecting that collection and taking chunks of size  λ and compressing them into a single field value by evaluating a polynomial. You do it repeatedly until the size of your output is  λ. 

If that's the case, then I don't think you need `ArrayChunk` trait, boxing iterators, recursion that require boxing, etc. I think you can achieve the same result by:

* Collecting your input into a vector of size `M`.  
* Iterating over chunks of size $\lambda$ and compressing them
* repeat `iteration` times


```rust
    let mut container = u_or_v_iterator.collect::<Vec<_>>();
    for i in 0..recursions_after_first {
        let mut write_pointer = 0;
        let step = if i == 0 { λ_FIRST } else { λ };

        for j in (0..container.len()).step_by(step) {
            let end = min(j + step, container.len());
            let chunk = &container[j..end];
            let compressed_val = if i == 0 {
                table_first.eval(chunk)[0]
            } else {
                tables[i - 1].eval(chunk)[0]
            };
            container[write_pointer] = compressed_val;
            write_pointer += 1;
        }

        container.truncate(write_pointer);
    }
```

this achieves the same goal with less complexity imo and I think it is faster. I am sure this can be optimized and a few abstractions may be required to make it look nicer, but the idea stays the same. And you don't need boxing and dynamic dispatch

cc: @danielmasny @benjaminsavage

Reply:
note that `let mut container = u_or_v_iterator.collect::<Vec<_>>();` can be avoided because you can just pass your container directly into this function to re-use the allocation. The fact that you have an iterator hints that you already have all these field values sitting somewhere in memory, so you can just re-use that

Reply:
I looked at the code again and seems that `BatchToVerify::verify` can take ownership of `u_and_v`, so the approach above could work

Reply:
@akoshelev Thanks for your suggestions! I think it is a memory cpu tradeoff. I think it would make sense to implement both versions and then simply benchmark them. This function should be very easy to benchmark since it is not even an async function. For the first iteration, we probably do not want to collect the u or v in a vector before compressing them since that would take around 670 Megabytes in memory for 50 million bit multiplications.